### PR TITLE
[REG 2.066] Issue 14838 - Wrong attribute inference for auto-generated class destructor with static array of non-POD type

### PR DESCRIPTION
**Pull request by 9rnsr**
https://issues.dlang.org/show_bug.cgi?id=14838

Add `_ArrayPostblit` and `_ArrayDtor`, then use them for static array copying and destruction.

### CODE REVIEW

File: src/module.c [hunk 1]
```diff
@@ -531,6 +531,69 @@ Module *Module::parse()
...
+            "bool _ArrayEq(T1, T2)(T1[] a, T2[] b) {\n"
+            " if (a.length != b.length) return false;\n"
+            " foreach (size_t i; 0 .. a.length) { if (a[i] != b[i]) return false; }\n"
+            " return true; }\n";
+
+        static const utf8_t code_ArrayPostblit[] =
+            "void _ArrayPostblit(T)(T[] a) { foreach (ref T e; a) e.__xpostblit(); }\n";
+
+        static const utf8_t code_ArrayDtor[] =
+            "void _ArrayDtor(T)(T[] a) { foreach_reverse (ref T e; a) e.__xdtor(); }\n";
```
Comment (MartinNowak) [inline src/module.c]: We already have those in druntime, called `_recursiveDestroy` and `_recursivePostblit`, but it's indeed an interesting idea to move them into the compiler.

**Comment by MartinNowak** [comment]
Mmh, seems like I shouldn't have spent my time on this (https://github.com/MartinNowak/dmd/commit/3a149d20ec44b2bb5eebf3b3a76f1d323adbf1d2).

File: src/module.c [hunk 2]
```diff
@@ -531,6 +531,69 @@ Module *Module::parse()
...
+            p.nextToken();
+            members->append(p.parseDeclDefs(0));
+        }
+        if (xopcmp)
+        {
+            Parser p(loc, this, code_xopCmp, strlen((const char *)code_xopCmp), 0);
+            p.nextToken();
+            members->append(p.parseDeclDefs(0));
+        }
+    }
```
Comment (MartinNowak) [inline src/module.c]: Why does it have to be `module object`? How about `module __internals` and then parsing a single piece of code.
That way you wouldn't need to deal with conflicting symbols.

File: src/interpret.c [hunk 3]
```diff
@@ -4759,48 +4759,41 @@ class Interpreter : public Visitor
...
+                    result = getVarExp(e->loc, istate, ((SymbolExp *)ea)->var, ctfeNeedRvalue);
+                else if (ea->op == TOKaddress)
+                    result = interpret(((AddrExp *)ea)->e1, istate);
+                else
+                    assert(0);
+                if (CTFEExp::isCantExp(result))
+                    return;
+
+                if (fd->ident == Id::_ArrayPostblit)
+                    result = evaluatePostblit(istate, result);
```
Comment (MartinNowak) [inline src/interpret.c]: Why is magic CTFE behavior for those necessary?

File: src/clone.c [hunk 4]
```diff
@@ -959,19 +983,36 @@ FuncDeclaration *buildDtor(AggregateDeclaration *ad, Scope *sc)
...
-            Expression *et = new TypeidExp(loc, v->type);
-            et = new DotIdExp(loc, et, Id::destroy);
-            ex = new CallExp(loc, et, ea);
+            // _ArrayDtor((cast(S*)this.v.ptr)[0 .. n])
+
+            // This is a hack so we can call destructors on const/immutable objects.
+            ex = new DotIdExp(loc, ex, Id::ptr);
+            ex = new CastExp(loc, ex, sdv->type->pointerTo());
+            if (stc & STCsafe)
+                stc = (stc & ~STCsafe) | STCtrusted;
```
Comment (MartinNowak) [inline src/clone.c]: Seems like this attribute change be visible on the type, e.g. `typeof(sd._xdtor)`?.
Maybe directly change `ex->type` instead of casting `.ptr` to avoid that

File: src/clone.c [hunk 5]
```diff
@@ -959,19 +983,36 @@ FuncDeclaration *buildDtor(AggregateDeclaration *ad, Scope *sc)
...
+            ex = new PtrExp(loc, ex);
+            if (stc & STCsafe)
+                stc = (stc & ~STCsafe) | STCtrusted;
+
             ex = new DotVarExp(loc, ex, sdv->dtor, 0);
             ex = new CallExp(loc, ex);
         }
         else
         {
-            // Typeinfo.destroy(cast(void*)&this.v);
```
Comment (MartinNowak) [inline src/clone.c]: You should assert here that `v->type->toBasetype()->ty == Tsarray` or use a switch with `default: assert(0);`.

File: src/clone.c [hunk 6]
```diff
@@ -822,19 +812,36 @@ FuncDeclaration *buildPostBlit(StructDeclaration *sd, Scope *sc)
         ex = new DotVarExp(loc, ex, v, 0);
         if (v->type->toBasetype()->ty == Tstruct)
         {
-            // this.v.__postblit()
+            // this.v.__xpostblit()
+
+            // This is a hack so we can call postblits on const/immutable objects.
+            ex = new AddrExp(loc, ex);
+            ex = new CastExp(loc, ex, v->type->mutableOf()->pointerTo());
+            ex = new PtrExp(loc, ex);
```
Comment (MartinNowak) [inline src/clone.c]: Why did that work before without the mutable cast?

File: src/e2ir.c [hunk 7]
```diff
@@ -795,12 +797,14 @@ elem *setArray(elem *eptr, elem *edim, Type *tb, elem *evalue, IRState *irs, int
...
-                {   /* Need to do postblit.
+                {
+                    /* Need to do postblit.
                      *   void *_d_arraysetassign(void *p, void *value, int dim, TypeInfo ti);
                      */
                     r = (op == TOKconstruct) ? RTLSYM_ARRAYSETCTOR : RTLSYM_ARRAYSETASSIGN;
                     evalue = el_una(OPaddr, TYnptr, evalue);
-                    elem *eti = getTypeInfo(tb, irs);
+                    // This is a hack so we can call postblits on const/immutable objects.
+                    elem *eti = getTypeInfo(tb->unSharedOf()->mutableOf(), irs);
```
Comment (MartinNowak) [inline src/e2ir.c]: Where does that come from, seems unrelated to this pull.

File: test/compilable/test14838.d [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+class C { A!int[1] array; }
+
+void test14838() pure nothrow @nogc @safe
+{
+    C c;
+    c.__xdtor();    // C.~this() will also be inferred to
+                    // pure nothrow @nogc @safe
+
+    A!int[1] array;
+    // scope destructor call does not cause attribute violation.
```
Comment (MartinNowak) [inline test/compilable/test14838.d]: I wonder why noone reported this earlier.

File: src/clone.c (hunk 5 above)
Comment (9rnsr) [inline src/clone.c]: It's not necessary. If `tv->ty` is `Tstruct`, `v->type->tobasetype()` is always either `Tstruct` or `Tsarray`. It's guaranteed by the `Type::baseElemOf()`  behavior.

File: src/clone.c [hunk 9]
```diff
@@ -959,19 +983,36 @@ FuncDeclaration *buildDtor(AggregateDeclaration *ad, Scope *sc)
...
-            ex = new CallExp(loc, et, ea);
+            // _ArrayDtor((cast(S*)this.v.ptr)[0 .. n])
+
+            // This is a hack so we can call destructors on const/immutable objects.
+            ex = new DotIdExp(loc, ex, Id::ptr);
+            ex = new CastExp(loc, ex, sdv->type->pointerTo());
+            if (stc & STCsafe)
+                stc = (stc & ~STCsafe) | STCtrusted;
+
+            uinteger_t n = v->type->size() / sdv->type->size();
```
Comment (MartinNowak) [inline src/clone.c]: Why not `((TypeSArray*)v)->dim->toInteger()` or simply a `SliceExp` without explicit bounds?

File: src/clone.c (hunk 5 above)
Comment (MartinNowak) [inline src/clone.c]: Yes, but these kind of implicit assumptions are hard to see and make it more difficult to understand or change the code, i.e. in order to read `buildDtor` I have to read `Type::baseElemOf`.

File: src/module.c (hunk 1 above)
Comment (9rnsr) [inline src/module.c]: Do you mean `_destructRecurse` and `_postblitRecurse`? I also planned them, but i decided not to use them.
Because they're specialized to `T[n]`. It will instantiate different templates for T[1], T[2], ..., so they would cause template bloat.

**Comment by MartinNowak** [comment]
Mostly looks good, thanks a lot.
I'm a bit unsure about parsing additional `object.d` members. A separate module might be better.

Will continue the review this evening.

File: src/module.c (hunk 2 above)
Comment (9rnsr) [inline src/module.c]: It would be better, but this is the most easy way that I can implement quickly. This change is for the regression issue, so I should not use risky way.

File: src/interpret.c (hunk 3 above)
Comment (9rnsr) [inline src/interpret.c]: Because CTFE does not support reinterpret cast from `T[n][m]` to `T*`.

File: src/module.c (hunk 2 above)
Comment (MartinNowak) [inline src/module.c]: A separate module might require linkage, but appending to object might also cause issues and sure is confusing. But you're right, let's do that as follow-up.

File: src/clone.c (hunk 4 above)
Comment (9rnsr) [inline src/clone.c]: When `ex->type` is set by hand, `ex->semantic()` will skip that sub-expressions. It's dangerous so it might send wrong AST to the backend.

File: src/clone.c (hunk 5 above)
Comment (9rnsr) [inline src/clone.c]: `Tstruct` and `Tsarray` are elaborate value types in D. So the use of `baseElemOf` to handle the two types is the common idiom in current dmd front end. (Actually `baseElemOf` function was added to reduce the boilerplate code.) Please learn the `baseElemOf` behavior.

File: src/clone.c (hunk 6 above)
Comment (9rnsr) [inline src/clone.c]: Maybe it didn't worked.

File: src/e2ir.c (hunk 7 above)
Comment (9rnsr) [inline src/e2ir.c]: Indeed it can be a small separated bug report, but essentially it's same with the hack in `buildPostblit`. If we don't fix this line, it will break `std.container.Array` unittest, because of the reference counting mismatch.

File: src/clone.c (hunk 9 above)
Comment (9rnsr) [inline src/clone.c]: Because it works even when `v->type` is multidimensional static array type (e.g. `T[2][3][4]`).

File: src/clone.c (hunk 5 above)
Comment (MartinNowak) [inline src/clone.c]: It's a simple matter of readability. Encoding all those implicit assumptions makes it much harder to understand, particularly for people less experienced with the compiler code.

```d
auto elemType = type->baseElemOf();
if (elemType != Tstruct)
    return;
if (type == Tstruct)
{
}
else
{
    // type has to be Tsarray b/c otherwise baseElemOf wouldn't be Tstruct
    // and baseElemOf does only peel of static arrays
}
```

vs.

```d
auto elemType = type->baseElemOf();
if (elemType != Tstruct)
    return;
switch (type)
{
case Tstruct:
    break;
case Tsarray:
    break;
default:
    assert(0);
}
```

Comment (MartinNowak) [inline src/clone.c]: And we force people to use final switch or switch with default in D for a good reason.

File: src/module.c (hunk 1 above)
Comment (MartinNowak) [inline src/module.c]: I agree, your solution is much nicer. Also we no longer need those functions in druntime, better if everything stays on one side.

File: src/clone.c (hunk 4 above)
Comment (MartinNowak) [inline src/clone.c]: I know, but taking the address will slightly change the type of the destructor/postblit, should be fine though.

```d
struct Bar
{
    ~this() @safe {}
}

struct Foo
{
    Bar[1] bar;
    ~this() @safe {}
}

pragma(msg, typeof(Foo.__xdtor));
```

File: src/declaration.c [hunk 10]
```diff
@@ -2063,36 +2063,42 @@ Expression *VarDeclaration::callScopeDtor(Scope *sc)
...
+            // v.__xdtor()
+            e = new VarExp(loc, this);
 
-                Expression *et = new TypeidExp(loc, type);
-                et = new DotIdExp(loc, et, Id::destroy);
+            /* This is a hack so we can call destructors on const/immutable objects.
+             * Need to add things like "const ~this()" and "immutable ~this()" to
+             * fix properly.
+             */
+            e->type = e->type->mutableOf();
```
Comment (MartinNowak) [inline src/declaration.c]: Here you are overwriting `e->type`. Why is that different from the case above?

Comment (9rnsr) [inline src/declaration.c]: Because `VarExp` is one of the leaf AST node. Its `semantic` will return itself. So the type setting is low risk.

**Comment by 9rnsr** [comment]
@MartinNowak Thanks!

**Comment by MartinNowak** [comment]
I've alreay seen a lot of error message à la `object._ArrayDtor isn't nothrow`.
This is somewhat misleading b/c object obviously doesn't have such a function and it's unclear who called `_ArrayDtor`. For the master branch we should establish an internal module, maybe we can also polish the error message to name the cause of the failure (dtor of whatever T isn't nothrow).

**Comment by MartinNowak** [comment]
Ping

**Comment by 9rnsr** [comment]
Internal module is not bad idea, but I'm not sure which name is the best for that module (`__builtin`, `core.compiler`, etc). Even if we reach to the best name, it's members are still invisible from dmd end users. So I'm not sure the approach will give benefits to users.

Better approach for the issue is to think about the ideal diagnostic message what you expects. Please give me some idea If you have.

**Comment by JackStouffer** [comment]
This introduced a regression https://issues.dlang.org/show_bug.cgi?id=15304
